fixes #7586 - user preferences for receiving mail notifications

stbenjam opened the pull request:
Adds a framework for user-selectable mail notifications.  The work is
still done in ActionMailer classes and launch by rake in cron, however a
wrapper called MailNotification is used to provide RBAC and make the
association with users.

Plugins may add their own mail notifications like they would add settings, see `mail_notification/foreman.rb`

stbenjam committed:
fixes #7586, #7734, #7172 - user preferences for receiving mail notifications

Adds a framework for user-selectable mail notifications.  The work is
still done in ActionMailer classes and launch by rake in cron, however a
wrapper called MailNotification is used to provide RBAC and make the
association with users.

Plugins may seed their own mail notifications, see
db/seeds.d/16-mail-notifications.rb for an example.

Authors:
  Shlomi Zadok <szadok@redhat.com>
  Stephen Benjamin <stephen@redhat.com>

shlomizadok commented on app/models/user.rb:
```diff
@@ -380,6 +400,11 @@ def prepare_password
 
+  def welcome_mail
+    return unless mail_enabled? && internal? && Setting[:send_welcome_email]
+    MailNotification[:welcome].deliver(:user => self, :password => password)
```
Not sure sending the password via email is so secure.

stbenjam replied:
I don't really care, I'm happy to remove it, I was just providing an example of how to do system mail notifications.

And generally, I agree for any other situation other than this one.  What's the argument against doing it here?  I did some research and found opinions both ways. 

The option is off by default, so it's opt-in. If an admin is creating a bunch of accounts for new users, it makes sense to be able to send a welcome mail like this to tell the user their new credentials, and a link to login -- they'll have to do this manually by hand, and it would probably be by mail or an internal chat system anyway.

I don't see this option as being really any different than say, sending a link with a token to login via e-mail.

shlomizadok commented on app/mailers/host_mailer.rb:
```diff
@@ -63,44 +73,19 @@ def summary(options = {})
-         :date => Time.now )
-  end
-
   def set_url
```
Should `set_url` and `load_metrics` be private?

domcleal commented:
Just glancing at the code, it needs better i18n support.  It's showing users descriptions and names of mail templates that don't appear to be extracted or translated.  (Is the point for this to be user-editable?  If not, why is it shown in the UI at all?)

The content of the mail is also translated in a few places, but it would appear to be in the locale of the person triggering the e-mail notification, rather than its recipient.

stbenjam commented:
>  It's showing users descriptions and names of mail templates that don't appear to be extracted or translated. (Is the point for this to be user-editable? If not, why is it shown in the UI at all?)

What do you mean by user editable? The names are shown to the user so they can select the notifications to receive.

I agree they should be translated though, thanks.

> The content of the mail is also translated in a few places, but it would appear to be in the locale of the person triggering the e-mail notification, rather than its recipient.

Would putting the `mail` call in the mailers in `User.as` block work, or would there be some better option to localize for the target receipient (e.g. an option to render)?

domcleal commented on config/routes.rb:
```diff
@@ -1,6 +1,12 @@
 require 'api_constraints'
 
 Foreman::Application.routes.draw do
+  resources :mail_notifications, :only => [:index, :show] do
```
Are these routes implemented?

domcleal commented on app/views/users/_form.html.erb:
```diff
@@ -35,6 +36,18 @@
+      <%= checkbox_f f, :mail_enabled %> 
+      <% if @user.new_record? %>
+        <div class="alert alert-block alert-warning">
+          <p><strong><%= _("Notice") %></strong> <%= _("Please save the user first before assigning mail notifications.") %></p>
```
This should really be avoided I think?

stbenjam replied:
Yes, but I don't want to show the full list, I want it to be confined to what the user has rights to, which means some javascript trickery + somehow using the auth system _before_ the user is saved, I was thinking that's  a lot of work for not very much gain.

domcleal commented on app/models/setting/puppet.rb:
```diff
@@ -12,7 +12,6 @@ def self.load_defaults
         self.set('document_root', N_("Document root where puppetdoc files should be created"), "#{Rails.root}/public/puppet/rdoc"),
         self.set('puppetrun', N_("Enable puppetrun support"), false),
         self.set('puppet_server', N_("Default Puppet server hostname"), "puppet"),
-        self.set('failed_report_email_notification', N_("Enable email alerts per each failed Puppet report"), false),
```
Missing DB migration for this?  It should also be removed from the DB.

stbenjam replied:
Done

domcleal commented on app/controllers/api/v2/mail_notifications_controller.rb:
```diff
@@ -0,0 +1,25 @@
+
+      before_filter :find_resource, :only => %w{show}
+
+      api :GET, "/mail_notifications/", N_("List of mail_notifications")
```
`s/_/ /`

domcleal commented:
> What do you mean by user editable? The names are shown to the user so they can select the notifications to receive.

I was looking at the routes for the UI thinking it might support editing of templates, but it doesn't, so that suggests they should be translated.  That said, the name is an API resource identifier, so I'm not sure that makes sense - that might need to change.

> Would putting the mail call in the mailers in User.as block work, or would there be some better option to localize for the target receipient (e.g. an option to render)?

I don't think User.as will set the locale, that's typically set via a controller filter based on headers + the user's preference, so you'll probably need to introduce a new helper and have a sane (English?) fallback.

domcleal commented:
It occurred to me last night that these resources are most similar to settings, and so perhaps model the i18n support on that.  They're not user provided or editable, but they're stored in the database by the application.  They should probably have static, untranslated names (identifiers), but translated descriptions which are used whenever possible?

stbenjam commented:
> It occurred to me last night that these resources are most similar to settings, and so perhaps model the i18n support on that. They're not user provided or editable, but they're stored in the database by the application. They should probably have static, untranslated names (identifiers), but translated descriptions which are used whenever possible?

That's great :+1:  The refactoring to make this work like Settings shouldn't be too bad.

shlomizadok commented:
[test]

stbenjam commented:
@shlomizadok If you want to have a look tomorrow - I changed the UI a little so we can show the now localized description.  Mail notifications now work exactly like Settings.  It may still need some improvements.

I was thinking that maybe 'alert' is a better name than 'error'? It's not really exposed to the user anyway, but for example a user is going to want an _alert_ that their host build completed.

Next step is to localize the e-mail content

stbenjam commented:
Descriptions are localized (~~except options - will fix~~)

![local9](https://cloud.githubusercontent.com/assets/429763/4679729/5c2695e8-5604-11e4-9763-2f37badc60ac.png)

Mail is localized:

![local](https://cloud.githubusercontent.com/assets/429763/4667954/6785063e-5561-11e4-86fe-5534644badfa.png)

@domcleal I implemented `mail_notifications` almost identically to settings, and I think it works OK, but it's not strictly necessary.  Seeding makes things simpler, the only problem is `rake gettext` doesn't look at `db` to get the strings, but it could right? Do you have a preference for either implementation?

domcleal commented:
@stbenjam I think extracting strings under db/ would be fine too

stbenjam commented:
Ok, moved back to seeds then

ohadlevy commented on app/mailers/application_mailer.rb:
```diff
@@ -0,0 +1,31 @@
+
+  private
+
+  class GroupMail
```
maybe I'm missing context, but why inner classes?

stbenjam replied:
It's only used below to provide the same interface to per-user localized group mail as the normal ActionMailer single mail object.

Would you put it somewhere else/do it a different way? In Java, I'd do it as an inner class.

ohadlevy replied:
I guess its fine, not that common.

usually one just either create a common method in the parent class or a service class... regardless, I'm fine with this.

ohadlevy commented on app/models/concerns/authorizable.rb:
```diff
@@ -6,20 +6,26 @@ module Authorizable
     }
 
+    def self.authorized(permission = nil, resource = nil)
+      self.authorized_as(User.current, permission, resource)
```
i didnt test, but once this is not a scope, does everything works as expected when you chain scopes? I assume yes as this method returns a scope..

stbenjam replied:
Yea, scopes work fine, e.g. something like `Host::Managed.authorized_as(user).with_os` or `Host::Managed.authorized(:view_hosts).with_os`  work fine.  Haven't seen anything odd with this, been running it for a few weeks

ohadlevy commented on app/models/mail_notification.rb:
```diff
@@ -0,0 +1,39 @@
+class MailNotification < ActiveRecord::Base
+  include Authorizable
+
+  INTERVALS = %w(Daily Weekly Monthly)
```
extract the strings?

ohadlevy commented on app/models/user.rb:
```diff
@@ -359,6 +383,11 @@ def prepare_password
 
+  def welcome_mail
+    return unless mail_enabled? && internal? && Setting[:send_welcome_email]
+    MailNotification[:welcome].deliver(:user => self)
```
nitpick: should we pass the user or the user_id? if later on we change this to async, passing the uid sounds better?

ohadlevy commented on app/models/user_mail_notification.rb:
```diff
@@ -0,0 +1,19 @@
+  belongs_to :user
+  belongs_to :mail_notification
+
+  validates :user, :presence => true
```
validates :used_id?

ohadlevy commented on app/models/usergroup.rb:
```diff
@@ -42,6 +42,10 @@ def recipients
   end
 
+  def recipients_for(notification)
+    all_users.collect { |user| user if user.receives?(notification) }.compact.sort.uniq
```
nitpick: can use select instead of collect and then you dont need compact.

ohadlevy commented on app/views/users/_form.html.erb:
```diff
@@ -35,6 +36,24 @@
+      <% end %>
+      <% if @user.new_record? %>
+        <div class="alert alert-block alert-warning">
+          <p><strong><%= _("Notice") %></strong> <%= _("Please save the user first before assigning mail notifications.") %></p>
```
why? nested_attributes should allow you to create everything in one go.

stbenjam replied:
@ohadlevy Ok, updated everything except for this one.

Like I mentioned on IRC, the main concern is the user's roles aren't saved yet, so it would probably need some javascript to update the mail notification page once roles are selected, and my JS  skills are pretty limited, so this would take me a while.

stbenjam commented:
Addressed all the feedback so far.

Sorry for pinging again, but I know you wanted this to be in at least a couple days before the RC.  @domcleal @ohadlevy @eLobato

dLobatog commented on test/fixtures/mail_notifications.yml:
```diff
@@ -0,0 +1,21 @@
+# Read about fixtures at http://ar.rubyonrails.org/classes/Fixtures.html
+puppet_summary:
```
I guess this is OK since this issue is urgent, but please file a refactor ticket that gets rid of this, we're trying to move to FactoryGirl.

stbenjam replied:
Right, I use factories most places, but the mail notifications have very specific names and purposes like Settings would -- you wouldn't just use a Factory to generate one, because the underlying mailer and method need to exist.

Fixtures are still appropriate there, right?

stbenjam replied:
(And note there is a MailNotification factory I use when the specific notification isn't relevant)

domcleal commented on app/views/users/_mail_notifications.html.erb:
```diff
@@ -0,0 +1,5 @@
+<%= f.hidden_field :mail_notification_id, :value => f.object.mail_notification.id %>
+<% values = f.object.mail_notification.subscription_type == 'alert' ? %w(Subscribe) : MailNotification::INTERVALS %>
```
'Subscribe' not extracted?

domcleal commented on app/mailers/host_mailer.rb:
```diff
@@ -1,47 +1,18 @@
-    end
-    email = options[:email] || Setting[:administrator]
-    raise ::Foreman::Exception.new(N_("unable to find recipients")) if email.empty?
+    raise ::Foreman::Exception.new(N_("Must specify a user with an e-mail")) unless (user=User.find(options[:user])) && user.mail
```
nitpick, no hyphen in Foreman strings, "email"

domcleal replied:
`user.mail_enabled?`?

dLobatog commented:
Merged as 3a36bdf6179cfb3207af623584c021addb674ffb , thanks @stbenjam !
